File: core/src/adictive/games/systems/RenderingSystem.java
package adictive.games.systems;

import com.badlogic.ashley.core.ComponentMapper;
import com.badlogic.ashley.core.Engine;
import com.badlogic.ashley.core.Entity;
import com.badlogic.ashley.core.EntityListener;
import com.badlogic.ashley.core.EntitySystem;
import com.badlogic.ashley.core.Family;
import com.badlogic.gdx.Gdx;
import com.badlogic.gdx.graphics.OrthographicCamera;
import com.badlogic.gdx.graphics.g2d.SpriteBatch;
import com.badlogic.gdx.graphics.g2d.TextureRegion;

import adictive.games.SquareWorld;
import adictive.games.components.PlayerComponent;
import adictive.games.components.TextureComponent;
import adictive.games.components.TransformComponent;
import adictive.games.utils.SortedOnInsertList;

public class RenderingSystem extends EntitySystem implements Reseteable {

    public static final int VIEWPORT_WIDTH_MTS = 15;
    public static final int VIEWPORT_HEIGHT_MTS = 15;
    public static final Family FAMILY = Family.all(TextureComponent.class, TransformComponent.class).get();

    private final SortedOnInsertList<Entity> renderQueue;

    private final ComponentMapper<TransformComponent> transformMapper;
    private final ComponentMapper<TextureComponent> textureMapper;
    private final SquareWorld world;
    private final SpriteBatch batch;

    public RenderingSystem(SquareWorld world) {
        super(10);
        this.world = world;
        this.batch = new SpriteBatch();
        textureMapper = ComponentMapper.getFor(TextureComponent.class);
        transformMapper = ComponentMapper.getFor(TransformComponent.class);

        renderQueue = new SortedOnInsertList<>(512,
                (entityA, entityB) ->
                        (int) Math.signum(transformMapper.get(entityB).pos.z - transformMapper.get(entityA).pos.z));
    }

    @Override
    public void reset() {
        renderQueue.clear();
    }

    @Override
    public void addedToEngine(Engine engine) {
        super.addedToEngine(engine);
        resize(world.getCamera(), Gdx.graphics.getWidth(), Gdx.graphics.getHeight());
        engine.addEntityListener(
                FAMILY,
                new EntityListener() {
                    @Override
                    public void entityAdded(Entity entity) {
                        renderQueue.add(entity);
                    }

                    @Override
                    public void entityRemoved(Entity entity) {
                        renderQueue.remove(entity);
                    }
                }
        );
    }

    @Override
    public void update(float deltaTime) {
        renderEntities();
    }

    private void renderEntity(Entity entity) {
        final TextureRegion textureRegion = textureMapper.get(entity).region;
        final TransformComponent transformComponent = transformMapper.get(entity);

        PlayerComponent pc = entity.getComponent(PlayerComponent.class);
        if (pc != null) {
            renderPlayerEffects(transformComponent, pc);
        }

        drawElement(
                transformComponent.pos.x,
                transformComponent.pos.y,
                textureRegion,
                transformComponent);
    }

    private void drawElement(float x, float y, TextureRegion textureRegion, TransformComponent transformComponent) {
        this.batch.draw(
                textureRegion,
                x,
                y,
                transformComponent.size.x / 2,
                transformComponent.size.y / 2,
                transformComponent.size.x,
                transformComponent.size.y,
                transformComponent.scale.x,
                transformComponent.scale.y,
                transformComponent.rotation
        );
    }

    private void renderPlayerEffects(TransformComponent tc, PlayerComponent pc) {
        drawElement(pc.trailX[1], pc.trailY[1], PlayerComponent.trail1, tc);
        drawElement(pc.trailX[2], pc.trailY[2], PlayerComponent.trail2, tc);
        drawElement(pc.trailX[3], pc.trailY[3], PlayerComponent.trail3, tc);
        drawElement(pc.trailX[4], pc.trailY[4], PlayerComponent.trail4, tc);
    }

    public void resize(int w, int h) {
        resize(world.getCamera(), w, h);
    }

    public static void resize(OrthographicCamera cam, int deviceWidth, int deviceHeight) {

        float width = deviceWidth / Gdx.graphics.getDensity();
        float height = deviceHeight / Gdx.graphics.getDensity();

        if (width < height) {
            cam.viewportWidth = VIEWPORT_WIDTH_MTS;
            cam.viewportHeight = VIEWPORT_HEIGHT_MTS * ((height) / width);
            cam.zoom = 200f / height;
        } else {
            cam.viewportWidth = VIEWPORT_WIDTH_MTS * ((width) / height);
            cam.viewportHeight = VIEWPORT_HEIGHT_MTS;
            cam.zoom = 400f / width;
        }
    }

    private void renderEntities() {
        world.getCamera().update();
        batch.setProjectionMatrix(world.getCamera().combined);

        batch.begin();
        for (Entity entity : renderQueue) {
            renderEntity(entity);
        }

        batch.end();
    }
}

File: core/src/adictive/games/systems/RenderingSystem.kt
package adictive.games.systems

import adictive.games.SquareWorld
import adictive.games.components.PlayerComponent
import adictive.games.components.TextureComponent
import adictive.games.components.TransformComponent
import adictive.games.utils.SortedOnInsertList
import com.badlogic.ashley.core.*
import com.badlogic.gdx.Gdx
import com.badlogic.gdx.graphics.OrthographicCamera
import com.badlogic.gdx.graphics.g2d.SpriteBatch
import com.badlogic.gdx.graphics.g2d.TextureRegion

class RenderingSystem(private val world: SquareWorld) : EntitySystem(10), Reseteable {

    private val renderQueue: SortedOnInsertList<Entity>

    private val transformMapper: ComponentMapper<TransformComponent>
    private val textureMapper: ComponentMapper<TextureComponent>
    private val batch: SpriteBatch

    init {
        this.batch = SpriteBatch()
        textureMapper = ComponentMapper.getFor(TextureComponent::class.java)
        transformMapper = ComponentMapper.getFor(TransformComponent::class.java)

        renderQueue = SortedOnInsertList(512,
                Comparator { entityA, entityB ->
                    Math.signum(
                            transformMapper.get(entityA).pos.z - transformMapper.get(entityB).pos.z
                    ).toInt()
                }
        )
    }

    override fun reset() {
        renderQueue.clear()
    }

    override fun addedToEngine(engine: Engine?) {
        super.addedToEngine(engine)
        resize(world.camera, Gdx.graphics.width, Gdx.graphics.height)
        engine!!.addEntityListener(
                FAMILY,
                object : EntityListener {
                    override fun entityAdded(entity: Entity) {
                        renderQueue.add(entity)
                    }

                    override fun entityRemoved(entity: Entity) {
                        renderQueue.remove(entity)
                    }
                }
        )
    }

    override fun update(deltaTime: Float) {
        renderEntities()
    }

    private fun renderEntity(entity: Entity) {
        val textureRegion = textureMapper.get(entity).region
        val transformComponent = transformMapper.get(entity)

        val pc = entity.getComponent<PlayerComponent>(PlayerComponent::class.java)
        if (pc != null) {
            renderPlayerEffects(transformComponent, pc)
        }

        drawElement(
                transformComponent.pos.x,
                transformComponent.pos.y,
                textureRegion,
                transformComponent)
    }

    private fun drawElement(x: Float, y: Float, textureRegion: TextureRegion?, transformComponent: TransformComponent) {
        this.batch.draw(
                textureRegion!!,
                x,
                y,
                transformComponent.size.x / 2,
                transformComponent.size.y / 2,
                transformComponent.size.x,
                transformComponent.size.y,
                transformComponent.scale.x,
                transformComponent.scale.y,
                transformComponent.rotation
        )
    }

    private fun renderPlayerEffects(tc: TransformComponent, pc: PlayerComponent) {
        drawElement(pc.trailX[1], pc.trailY[1], PlayerComponent.trail1, tc)
        drawElement(pc.trailX[2], pc.trailY[2], PlayerComponent.trail2, tc)
        drawElement(pc.trailX[3], pc.trailY[3], PlayerComponent.trail3, tc)
        drawElement(pc.trailX[4], pc.trailY[4], PlayerComponent.trail4, tc)
    }

    fun resize(w: Int, h: Int) {
        resize(world.camera, w, h)
    }

    private fun renderEntities() {
        world.camera.update()
        batch.projectionMatrix = world.camera.combined

        batch.begin()
        for (entity in renderQueue) {
            renderEntity(entity)
        }

        batch.end()
    }

    companion object {

        const val VIEWPORT_WIDTH_MTS = 15
        const val VIEWPORT_HEIGHT_MTS = 15
        val FAMILY = Family.all(TextureComponent::class.java, TransformComponent::class.java).get()

        fun resize(cam: OrthographicCamera, deviceWidth: Int, deviceHeight: Int) {

            val width = deviceWidth / Gdx.graphics.density
            val height = deviceHeight / Gdx.graphics.density

            if (width < height) {
                cam.viewportWidth = VIEWPORT_WIDTH_MTS.toFloat()
                cam.viewportHeight = VIEWPORT_HEIGHT_MTS * (height / width)
                cam.zoom = 200f / height
            } else {
                cam.viewportWidth = VIEWPORT_WIDTH_MTS * (width / height)
                cam.viewportHeight = VIEWPORT_HEIGHT_MTS.toFloat()
                cam.zoom = 400f / width
            }
        }
    }
}
